keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,344 @@
+[`kubernetes.io/os`](https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetes-io-os) to identify
+Windows nodes from their Linux counterparts. If the RuntimClass's scheduling field has a nodeSelector  `kubernetes.io/os: 'windows'`,
+we can authoritatively say that the pod is indeed a Windows one during the api-server admission time. Querying the RuntimeClass
+during admission time may be expensive operation as the admission plugin is not stateless anymore. 
+
+### Risks and Mitigations
+
+
+
+Existing users may be impacted as we're enforcing to use RuntimeClasses instead of plain nodeSelector and tolerations.
+
+
+The downside of this approach is since we're making RuntimeClass as default choice for the users, existing users have to update their pod spec or workload pod template spec to ensure that their workloads are not broken during upgrades. 
+This can be mitigated by 
+- Having appropriate warnings in the releases before we beta and GA will mitigate the problem.
+- Having an example out-of-tree admission webhook which can detect such workloads and update them.(Having an in-tree admission plugin which mutates the pod spec to have appropriate Windows specific RuntimeClass may be too hard to do generically and not transparent to the user)
+
+
+## Design Details
+
+
+We can piggyback on the existing RuntimeClass admission controller to query for the RuntimeClass and see if it has
+`kubernetes.io/os: windows` in it's scheduling field. 
+TODO: We need to get an ack from sig-auth to use it.
+
+If we cannot piggyback on the RuntimeClass controller we need to build our own validating admission controller for Windows specific pods.
+
+### Test Plan
+
+
+Whatever path we choose, unit and e2e tests are a hard requirement to ensure smooth
+transition to make RuntimeClasses default choice for the users.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+- Users getting a warning to switch to RuntimeClasses in beta release
+
+#### Alpha -> Beta Graduation
+- Gather feedback from end users
+- Tests are in Testgrid and linked in the KEP
+- Deny admission to pods that have only `kubernetes.io/os: windows` in nodeSelector
+  but don't have an associated RuntimeClass with `kubernetes.io/os: windows` in
+  scheduling field.
+
+#### Alpha -> Beta Graduation
+- 2 examples of end users using this field
+
+### Upgrade / Downgrade Strategy
+
+- Upgrades:
+  When upgrading from a release without this feature, to a release with `RuntimeClassesForWindows` enabled, we will honor both nodeSelector and RuntimeClasses
+  but a warning will be sent to the user that `honoring explicit nodeSelector without Runtimeclasses will be stopped from next release` if the pod has just nodeSelector set. This ensures users have enough time to move without breaking their workloads.
```

Comment:
Please call out how this warning will be sent. Event etc.

Reply:
It'll be sent as a message to the user before [persisting](https://github.com/kubernetes/kubernetes/blob/c2882f314218af489267b7163b2ca09c79ffb00a/pkg/registry/apps/daemonset/strategy.go#L170)